Interesting, I still have the same issue with your patch:

```
(venv) milan@mars:~/Work/neural-fortran/neural-fortran_scivision/build$ cmake .. -DSERIAL=1
-- The C compiler identification is GNU 9.4.0
-- The Fortran compiler identification is GNU 9.4.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting Fortran compiler ABI info
-- Detecting Fortran compiler ABI info - done
-- Check for working Fortran compiler: /usr/bin/gfortran - skipped
-- Configuring build for serial execution
-- h5fortran 4.6.3  CMake 3.22.5
-- checking that C and Fortran compilers can link
-- checking that C and Fortran compilers can link - OK
-- Looking for H5_HAVE_FILTER_SZIP
-- Looking for H5_HAVE_FILTER_SZIP - found
-- Looking for H5_HAVE_FILTER_DEFLATE
-- Looking for H5_HAVE_FILTER_DEFLATE - found
-- Looking for H5_HAVE_PARALLEL
-- Looking for H5_HAVE_PARALLEL - not found
-- Found ZLIB: /usr/lib/x86_64-linux-gnu/libz.so (found version "1.2.11") 
-- Found SZIP: /usr/lib/x86_64-linux-gnu/libsz.so  
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Performing Test HDF5_C_links
-- Performing Test HDF5_C_links - Success
-- Performing Test HDF5_Fortran_links
-- Performing Test HDF5_Fortran_links - Success
-- Found HDF5: /usr/lib/x86_64-linux-gnu/hdf5/serial/lib/libhdf5_hl.so;/usr/lib/x86_64-linux-gnu/hdf5/serial/lib/libhdf5.so (found suitable version "1.10.4", minimum required is "1.8.7") found components: Fortran HL 
CMake Deprecation Warning at build/_deps/jsonfortran-src/CMakeLists.txt:12 (cmake_minimum_required):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- CMake build configuration for JSON-Fortran 8.3.0
-- Dynamically computing FORD output information...
-- Running FORD to dynamically compute documentation outputs, this could take a while...
-- Caching FORD outputs
-- Done dynamically computing FORD outputs.
-- Found HDF5: /usr/lib/x86_64-linux-gnu/hdf5/serial/lib/libhdf5_hl.so;/usr/lib/x86_64-linux-gnu/hdf5/serial/lib/libhdf5.so (found version "1.10.4") found components: Fortran 
-- Configuring to build with -fcoarray=single
-- Configuring done
-- Generating done
-- Build files have been written to: /home/milan/Work/neural-fortran/neural-fortran_scivision/build
(venv) milan@mars:~/Work/neural-fortran/neural-fortran_scivision/build$ make
Scanning dependencies of target neural
[  0%] Building Fortran object CMakeFiles/neural.dir/src/nf/nf_datasets_mnist.f90.o
[  2%] Building Fortran object CMakeFiles/neural.dir/src/nf/nf_activation.f90.o
[  2%] Building Fortran object CMakeFiles/neural.dir/src/nf/nf_base_layer.f90.o
[  4%] Building Fortran object CMakeFiles/neural.dir/src/nf/nf_layer.f90.o
[  6%] Building Fortran object CMakeFiles/neural.dir/src/nf/nf_layer_constructors.f90.o
[  8%] Building Fortran object CMakeFiles/neural.dir/src/nf/nf_optimizers.f90.o
[ 10%] Building Fortran object CMakeFiles/neural.dir/src/nf/nf_network.f90.o
[ 10%] Building Fortran object CMakeFiles/neural.dir/src/nf.f90.o
[ 10%] Building Fortran object CMakeFiles/neural.dir/src/nf/io/nf_io_binary.f90.o
[ 12%] Building Fortran object CMakeFiles/neural.dir/src/nf/io/nf_io_binary_submodule.f90.o
[ 12%] Building Fortran object CMakeFiles/neural.dir/src/nf/io/nf_io_hdf5.f90.o
[ 14%] Building Fortran object CMakeFiles/neural.dir/src/nf/io/nf_io_hdf5_submodule.f90.o
f951: Fatal Error: Module file ‘h5fortran.mod’ has not been generated, either because the module does not contain a MODULE PROCEDURE or there is an error in the module.
compilation terminated.
make[2]: *** [CMakeFiles/neural.dir/build.make:569: CMakeFiles/neural.dir/src/nf/io/nf_io_hdf5_submodule.f90.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:251: CMakeFiles/neural.dir/all] Error 2
make: *** [Makefile:166: all] Error 2
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -12,96 +12,13 @@ LANGUAGES C Fortran
 
 enable_testing()
 
-option(SERIAL "Serial execution")
-
-# Set output paths for modules, archives, and executables
-set(CMAKE_Fortran_MODULE_DIRECTORY ${PROJECT_BINARY_DIR}/include)
-set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
-set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
-set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
-
-if(SERIAL)
-  message(STATUS "Configuring build for serial execution")
-else()
-  message(STATUS "Configuring build for parallel execution")
-endif()
-
 include(FetchContent)
 
-set(h5fortran_BUILD_TESTING false)
+include(cmake/options.cmake)
+include(cmake/compilers.cmake)
 
-FetchContent_Declare(
-  h5fortran
-  GIT_REPOSITORY https://github.com/geospace-code/h5fortran
-  GIT_TAG v4.6.3
-  GIT_SHALLOW true
-)
-
-FetchContent_Declare(
-  jsonfortran
-  GIT_REPOSITORY https://github.com/jacobwilliams/json-fortran
-  GIT_TAG 8.3.0
-  GIT_SHALLOW true
-)
-
-FetchContent_MakeAvailable(h5fortran jsonfortran)
-
-file(MAKE_DIRECTORY ${h5fortran_BINARY_DIR}/include)
-include_directories(${h5fortran_BINARY_DIR}/include)
-
-file(MAKE_DIRECTORY ${jsonfortran_BINARY_DIR}/include)
-include_directories(${jsonfortran_BINARY_DIR})
-
-list(APPEND CMAKE_MODULE_PATH ${h5fortran_SOURCE_DIR}/cmake/Modules)
-find_package(HDF5 COMPONENTS Fortran REQUIRED)
-
-# compiler flags for gfortran
-if(CMAKE_Fortran_COMPILER_ID STREQUAL "GNU")
-
-  if(SERIAL)
-    message(STATUS "Configuring to build with -fcoarray=single")
-    add_compile_options("$<$<COMPILE_LANGUAGE:Fortran>:-fcoarray=single>")
-  else()
-    add_compile_options("$<$<COMPILE_LANGUAGE:Fortran>:-fcoarray=lib>")
-  endif()
-
-  if(BLAS)
-    add_compile_options("$<$<COMPILE_LANGUAGE:Fortran>:-fexternal-blas;${BLAS}>")
-    list(APPEND LIBS "blas")
-    message(STATUS "Configuring build to use BLAS from ${BLAS}")
-  endif()
-
-  add_compile_options("$<$<AND:$<COMPILE_LANGUAGE:Fortran>,$<CONFIG:Debug>>:-fcheck=bounds;-fbacktrace>")
-  add_compile_options("$<$<AND:$<COMPILE_LANGUAGE:Fortran>,$<CONFIG:Release>>:-Ofast;-fno-frontend-optimize;-fno-backtrace>")
-
-elseif(CMAKE_Fortran_COMPILER_ID MATCHES "^Intel")
-  # compiler flags for ifort
-
-  if(SERIAL)
-    message(STATUS "Configuring to build with -coarray=single")
-    if(WIN32)
-      add_compile_options("$<$<COMPILE_LANGUAGE:Fortran>:/Qcoarray:single>")
-    else()
-      add_compile_options("$<$<COMPILE_LANGUAGE:Fortran>:-coarray=single>")
-    endif()
-  else()
-    if(WIN32)
-      add_compile_options("$<$<COMPILE_LANGUAGE:Fortran>:/Qcoarray:shared>")
-    else()
-      add_compile_options("$<$<COMPILE_LANGUAGE:Fortran>:-coarray=shared>")
-    endif()
-  endif()
-
-  string(APPEND CMAKE_Fortran_FLAGS " -assume byterecl")
-  add_compile_options("$<$<AND:$<COMPILE_LANGUAGE:Fortran>,$<CONFIG:Debug>>:-check;-traceback>")
-  # add_compile_options("$<$<AND:$<COMPILE_LANGUAGE:Fortran>,$<CONFIG:Release>>:-O3>")
-
-elseif(CMAKE_Fortran_COMPILER_ID STREQUAL "Cray")
-  # compiler flags for Cray ftn
-  string(APPEND CMAKE_Fortran_FLAGS " -h noomp")
-  add_compile_options("$<$<AND:$<COMPILE_LANGUAGE:Fortran>,$<CONFIG:Debug>>:-O0;-g>")
-  add_compile_options("$<$<AND:$<COMPILE_LANGUAGE:Fortran>,$<CONFIG:Release>>:-O3>")
-endif()
+include(cmake/h5fortran.cmake)
+include(cmake/json.cmake)
 
 # library to archive (libneural.a)
 add_library(neural
@@ -145,8 +62,9 @@ add_library(neural
   src/nf/io/nf_io_hdf5.f90
   src/nf/io/nf_io_hdf5_submodule.f90
 )
-target_link_libraries(neural PRIVATE h5fortran::h5fortran HDF5::HDF5)
-target_include_directories(neural PRIVATE ${jsonfortran_BINARY_DIR}/include)
+target_link_libraries(neural PRIVATE h5fortran::h5fortran HDF5::HDF5 jsonfortran::jsonfortran)
+
+install(TARGETS neural)
 
 # Remove leading or trailing whitespace
 string(REGEX REPLACE "^ | $" "" LIBS "${LIBS}")
@@ -155,11 +73,11 @@ string(REGEX REPLACE "^ | $" "" LIBS "${LIBS}")
 
 foreach(execid input1d_layer input3d_layer dense_layer conv2d_layer maxpool2d_layer flatten_layer dense_network dense_network_from_keras conv2d_network io_hdf5 keras_read_model)
   add_executable(test_${execid} test/test_${execid}.f90)
-  target_link_libraries(test_${execid} PRIVATE neural h5fortran::h5fortran jsonfortran ${LIBS})
+  target_link_libraries(test_${execid} PRIVATE neural h5fortran::h5fortran jsonfortran::jsonfortran ${LIBS})
   add_test(test_${execid} bin/test_${execid})
 endforeach()
 
 foreach(execid cnn mnist mnist_from_keras simple sine)
   add_executable(${execid} example/${execid}.f90)
-  target_link_libraries(${execid} PRIVATE neural h5fortran::h5fortran jsonfortran ${LIBS})
+  target_link_libraries(${execid} PRIVATE neural h5fortran::h5fortran jsonfortran::jsonfortran ${LIBS})
 endforeach()

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -34,12 +34,14 @@ FetchContent_Declare(
   h5fortran
   GIT_REPOSITORY https://github.com/geospace-code/h5fortran
   GIT_TAG v4.6.3
+  GIT_SHALLOW true
 )
 
 FetchContent_Declare(
   jsonfortran
   GIT_REPOSITORY https://github.com/jacobwilliams/json-fortran
   GIT_TAG 8.3.0
+  GIT_SHALLOW true
 )
 
 FetchContent_MakeAvailable(h5fortran jsonfortran)

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -143,7 +143,7 @@ add_library(neural
   src/nf/io/nf_io_hdf5.f90
   src/nf/io/nf_io_hdf5_submodule.f90
 )
-target_link_libraries(neural PRIVATE HDF5::HDF5)
+target_link_libraries(neural PRIVATE h5fortran::h5fortran HDF5::HDF5)
 target_include_directories(neural PRIVATE ${jsonfortran_BINARY_DIR}/include)
 
 # Remove leading or trailing whitespace

--- a/cmake/compilers.cmake
+++ b/cmake/compilers.cmake
@@ -1,0 +1,51 @@
+# compiler flags for gfortran
+if(CMAKE_Fortran_COMPILER_ID STREQUAL "GNU")
+
+  if(SERIAL)
+    message(STATUS "Configuring to build with -fcoarray=single")
+    add_compile_options("$<$<COMPILE_LANGUAGE:Fortran>:-fcoarray=single>")
+  else()
+    add_compile_options("$<$<COMPILE_LANGUAGE:Fortran>:-fcoarray=lib>")
+  endif()
+
+  if(BLAS)
+    add_compile_options("$<$<COMPILE_LANGUAGE:Fortran>:-fexternal-blas;${BLAS}>")
+    list(APPEND LIBS "blas")
+    message(STATUS "Configuring build to use BLAS from ${BLAS}")
+  endif()
+
+  add_compile_options("$<$<AND:$<COMPILE_LANGUAGE:Fortran>,$<CONFIG:Debug>>:-fcheck=bounds;-fbacktrace>")
+  add_compile_options("$<$<AND:$<COMPILE_LANGUAGE:Fortran>,$<CONFIG:Release>>:-Ofast;-fno-frontend-optimize;-fno-backtrace>")
+
+elseif(CMAKE_Fortran_COMPILER_ID MATCHES "^Intel")
+  # compiler flags for ifort
+
+  if(SERIAL)
+    message(STATUS "Configuring to build with -coarray=single")
+    if(WIN32)
+      add_compile_options("$<$<COMPILE_LANGUAGE:Fortran>:/Qcoarray:single>")
+    else()
+      add_compile_options("$<$<COMPILE_LANGUAGE:Fortran>:-coarray=single>")
+    endif()
+  else()
+    if(WIN32)
+      add_compile_options("$<$<COMPILE_LANGUAGE:Fortran>:/Qcoarray:shared>")
+    else()
+      add_compile_options("$<$<COMPILE_LANGUAGE:Fortran>:-coarray=shared>")
+    endif()
+  endif()
+
+  if(WIN32)
+    string(APPEND CMAKE_Fortran_FLAGS " /assume:byterecl")
+  else()
+    string(APPEND CMAKE_Fortran_FLAGS " -assume byterecl")
+  endif()
+  add_compile_options("$<$<AND:$<COMPILE_LANGUAGE:Fortran>,$<CONFIG:Debug>>:-check;-traceback>")
+  # add_compile_options("$<$<AND:$<COMPILE_LANGUAGE:Fortran>,$<CONFIG:Release>>:-O3>")
+
+elseif(CMAKE_Fortran_COMPILER_ID STREQUAL "Cray")
+  # compiler flags for Cray ftn
+  string(APPEND CMAKE_Fortran_FLAGS " -h noomp")
+  add_compile_options("$<$<AND:$<COMPILE_LANGUAGE:Fortran>,$<CONFIG:Debug>>:-O0;-g>")
+  add_compile_options("$<$<AND:$<COMPILE_LANGUAGE:Fortran>,$<CONFIG:Release>>:-O3>")
+endif()

--- a/cmake/h5fortran.cmake
+++ b/cmake/h5fortran.cmake
@@ -1,0 +1,15 @@
+set(h5fortran_BUILD_TESTING false)
+
+FetchContent_Declare(h5fortran
+  GIT_REPOSITORY https://github.com/geospace-code/h5fortran
+  GIT_TAG v4.6.3
+  GIT_SHALLOW true
+)
+
+FetchContent_MakeAvailable(h5fortran)
+
+file(MAKE_DIRECTORY ${h5fortran_BINARY_DIR}/include)
+
+
+list(APPEND CMAKE_MODULE_PATH ${h5fortran_SOURCE_DIR}/cmake/Modules)
+find_package(HDF5 COMPONENTS Fortran REQUIRED)

--- a/cmake/json.cmake
+++ b/cmake/json.cmake
@@ -1,0 +1,35 @@
+# use our own CMake script to build jsonfortran instead of jsonfortran/CMakelists.txt
+
+FetchContent_Declare(jsonfortran
+  GIT_REPOSITORY https://github.com/jacobwilliams/json-fortran
+  GIT_TAG 8.3.0
+  GIT_SHALLOW true
+)
+
+FetchContent_Populate(jsonfortran)
+
+SET(JSON_REAL_KIND "REAL64")
+SET(JSON_INT_KIND "INT32")
+
+set(_src ${jsonfortran_SOURCE_DIR}/src)
+
+set (JF_LIB_SRCS
+${_src}/json_kinds.F90
+${_src}/json_parameters.F90
+${_src}/json_string_utilities.F90
+${_src}/json_value_module.F90
+${_src}/json_file_module.F90
+${_src}/json_module.F90
+)
+
+add_library(jsonfortran ${JF_LIB_SRCS})
+target_compile_definitions(jsonfortran PRIVATE ${JSON_REAL_KIND} ${JSON_INT_KIND})
+target_include_directories(jsonfortran PUBLIC
+$<BUILD_INTERFACE:${CMAKE_BINARY_DIR}/include>
+$<INSTALL_INTERFACE:include>
+)
+
+add_library(jsonfortran::jsonfortran INTERFACE IMPORTED GLOBAL)
+target_link_libraries(jsonfortran::jsonfortran INTERFACE jsonfortran)
+
+install(TARGETS jsonfortran)

--- a/cmake/options.cmake
+++ b/cmake/options.cmake
@@ -1,0 +1,13 @@
+option(SERIAL "Serial execution")
+
+# Set output paths for modules, archives, and executables
+set(CMAKE_Fortran_MODULE_DIRECTORY ${CMAKE_BINARY_DIR}/include)
+set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
+set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
+set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
+
+if(SERIAL)
+  message(STATUS "Configuring build for serial execution")
+else()
+  message(STATUS "Configuring build for parallel execution")
+endif()

